File: library/src/main/java/com/zhanghao/log/LogPrintImpl.java
package com.zhanghao.log;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


/**
 * Created by zhanghao.
 * 2017/10/8-13:52
 */

public class LogPrintImpl implements LogPrint {

    private boolean isLogToAndroid = true;
    private boolean isLogToDisk;

    private OutputAndroidLogcat outputAndroidLogcat;
    private OutputDisk outputDisk;

    private String diskFilePath;
    private int diskFileMaxSize;
    private OutputAnyAdapter outputAnyAdapter;

    private static LogPrintImpl logPrintImpl;
    /**
     * Priority constant for the println method; use Log.v.
     */
    public static final int VERBOSE = 2;

    /**
     * Priority constant for the println method; use Log.d.
     */
    public static final int DEBUG = 3;

    /**
     * Priority constant for the println method; use Log.i.
     */
    public static final int INFO = 4;

    /**
     * Priority constant for the println method; use Log.w.
     */
    public static final int WARN = 5;

    /**
     * Priority constant for the println method; use Log.e.
     */
    public static final int ERROR = 6;

    /**
     * Priority constant for the println method.
     */
    public static final int ASSERT = 7;

    private Handler handler = new Handler(Looper.getMainLooper());
    /**
     * log线程池
     */
    ExecutorService logThreadPool;

    private LogPrintImpl() {
        outputAndroidLogcat = new OutputAndroidLogcat();
    }

    static LogPrintImpl getInstance() {
        if (logPrintImpl == null) {
            synchronized (LogPrintImpl.class) {
                if (logPrintImpl == null) {
                    logPrintImpl = new LogPrintImpl();
                }
            }
        }
        return logPrintImpl;
    }

    public void init(Builder builder) {
        this.isLogToAndroid = builder.isLogToAndroid;
        this.isLogToDisk = builder.isLogToDisk;
        this.diskFilePath = builder.diskFilePath;
        this.diskFileMaxSize = builder.diskFileMaxSize;
        //文件目录不存在，则不会创建输出到sd卡的日志
        if(diskFilePath != null) {
            this.outputDisk = new OutputDisk(diskFilePath, diskFileMaxSize);
        }
        this.outputAnyAdapter = builder.outputAdapter;
    }

    @Override
    public void d(String tag, String message, String keyword) {
        LogBean logBean = createLogBean(DEBUG, tag, message, keyword);
        printLog(logBean);
    }

    @Override
    public void e(String tag, String message, String keyword) {
        LogBean logBean = createLogBean(ERROR, tag, message, keyword);
        printLog(logBean);
    }

    @Override
    public void e(String tag, Throwable throwable) {
        e(tag, throwable, null);
    }

    @Override
    public void e(String tag, String message, Throwable throwable) {
        e(tag, message, throwable, null);
    }


    @Override
    public void e(String tag, Throwable throwable, String keyword) {
        e(tag, null, throwable, keyword);
    }

    @Override
    public void e(String tag, String message, Throwable throwable, String keyword) {
        if (throwable != null && message != null) {
            message += " : " + Utils.getStackTraceString(throwable);
        }
        if (throwable != null && message == null) {
            message = Utils.getStackTraceString(throwable);
        }
        if (Utils.isEmpty(message)) {
            message = "Empty/NULL log message";
        }
        LogBean logBean = createLogBean(ERROR, tag, message, keyword);
        printLog(logBean);
    }

    @Override
    public void i(String tag, String message, String keyword) {
        LogBean logBean = createLogBean(INFO, tag, message, keyword);
        printLog(logBean);
    }

    @Override
    public void w(String tag, String message, String keyword) {
        LogBean logBean = createLogBean(WARN, tag, message, keyword);
        printLog(logBean);
    }

    @Override
    public void v(String tag, String message, String keyword) {
        LogBean logBean = createLogBean(VERBOSE, tag, message, keyword);
        printLog(logBean);
    }

    @Override
    public void d(String tag, String message) {
        LogBean logBean = createLogBean(DEBUG, tag, message, null);
        printLog(logBean);
    }

    @Override
    public void e(String tag, String message) {
        LogBean logBean = createLogBean(ERROR, tag, message, null);
        printLog(logBean);
    }

    @Override
    public void i(String tag, String message) {
        LogBean logBean = createLogBean(INFO, tag, message, null);
        printLog(logBean);
    }

    @Override
    public void w(String tag, String message) {
        LogBean logBean = createLogBean(WARN, tag, message, null);
        printLog(logBean);
    }

    @Override
    public void v(String tag, String message) {
        LogBean logBean = createLogBean(VERBOSE, tag, message, null);
        printLog(logBean);
    }

    @Override
    public void w(String message) {
        w(null, message);
    }

    @Override
    public void v(String message) {
        v(null, message);
    }

    private LogBean createLogBean(int debug, String tag, String message, String keyword) {
        LogBean logBean = new LogBean();
        logBean.setPriority(debug);
        logBean.setTag(tag);
        logBean.setStackTraceElement(Thread.currentThread().getStackTrace());
        logBean.setTime(System.currentTimeMillis());
        logBean.setContent(message);
        logBean.setKeyword(keyword);
        return logBean;
    }

    public void setOutputAndroidLogcat(OutputAndroidLogcat outputAndroidLogcat) {
        this.outputAndroidLogcat = outputAndroidLogcat;
    }

    public void setOutputDisk(OutputDisk outputDisk) {
        this.outputDisk = outputDisk;
    }

    public void setOutputAdapter(OutputAnyAdapter logAdapter) {
        this.outputAnyAdapter = logAdapter;
    }

    public void setLogToAndroid(boolean logToAndroid) {
        isLogToAndroid = logToAndroid;
    }

    public void setLogToDisk(boolean logToDisk) {
        isLogToDisk = logToDisk;
    }

    /**
     * 设置线程池
     * @param threadPool 线程池用于打印日志
     */
    public void setThreadPool(ExecutorService threadPool) {
        this.logThreadPool = threadPool;
    }

    /**
     * 获取线程池对象
     * @return 线程池用于打印日志
     */
    public ExecutorService getThreadPool() {
        return this.logThreadPool;
    }

    /**
     * 关闭线程池
     */
    public void shutdownExecutorService() {
        this.logThreadPool.shutdown();
    }

    /**
     * 打印日志
     * @param logBean 日志对象
     * 用固定线程池可保证日志按顺序打印,扩展日志打印需要用户自己指定线程
     */
    private void printLog(final LogBean logBean) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                try {
                    if(logThreadPool == null){
                        logThreadPool = Executors.newFixedThreadPool(1);
                    }
                    logThreadPool.execute(new Runnable() {
                        @Override
                        public void run() {
                            try {
                                if (isLogToAndroid) {
                                    outputAndroidLogcat.log((LogBean) logBean.clone());
                                }
                                if (isLogToDisk) {
                                    if (outputDisk != null){
                                        outputDisk.log((LogBean) logBean.clone());
                                    }
                                }
                            } catch (CloneNotSupportedException e) {
                                e.printStackTrace();
                            }
                        }
                    });
                    if (outputAnyAdapter != null) {
                        outputAnyAdapter.log((LogBean) logBean.clone());
                    }
                } catch (CloneNotSupportedException e) {
                    e.printStackTrace();
                }
            }
        });
    }

    @Override
    public void d(String message) {
        d(null, message);
    }

    @Override
    public void e(String message) {
        e(null, message);
    }

    @Override
    public void i(String message) {
        i(null, message);
    }

    public static final class Builder {

        private String diskFilePath;
        private int diskFileMaxSize;
        private boolean isLogToAndroid = true;
        private boolean isLogToDisk;

        private OutputAnyAdapter outputAdapter;

        public Builder setDiskFilePath(String diskFilePath) {
            this.diskFilePath = diskFilePath;
            return this;
        }

        public Builder setLogOpen(boolean isOpen) {
            isLogToAndroid = isOpen;
            isLogToDisk = isOpen;
            return this;
        }

        public Builder setDiskFileMaxSize(int diskFileMaxSize) {
            this.diskFileMaxSize = diskFileMaxSize;
            return this;
        }

        public Builder setLogToAndroid(boolean logToAndroid) {
            isLogToAndroid = logToAndroid;
            return this;
        }

        public Builder setLogToDisk(boolean logToDisk) {
            isLogToDisk = logToDisk;
            return this;
        }

        public Builder setOutputAdapter(OutputAnyAdapter outputAdapter) {
            this.outputAdapter = outputAdapter;
            return this;
        }

    }
}

File: library/src/main/java/com/zhanghao/log/LogPrintImpl.kt
package com.zhanghao.log

import android.os.Handler
import android.os.Looper
import java.io.FileNotFoundException
import java.lang.Exception


/**
 * Created by zhanghao.
 * 2017/10/8-13:52
 */

class LogPrintImpl private constructor() : LogPrint {

    private var isLogToAndroid = true
    private var isLogToDisk: Boolean = false

    private var outputAndroidLogcat: OutputAndroidLogcat? = null
    private var outputDisk: OutputDisk? = null

    private var diskFilePath: String? = null
    private var diskFileMaxSize: Int = 0
    private var outputAnyAdapter: OutputAnyAdapter? = null

    private val handler = Handler(Looper.getMainLooper())

    init {
        outputAndroidLogcat = OutputAndroidLogcat()
    }

    fun init(builder: Builder) {
        this.isLogToAndroid = builder.isLogToAndroid
        this.isLogToDisk = builder.isLogToDisk
        this.diskFilePath = builder.diskFilePath
        this.diskFileMaxSize = builder.diskFileMaxSize
        //文件目录不存在，则不会创建输出到sd卡的日志
        if (diskFilePath != null) {
            this.outputDisk = OutputDisk(diskFilePath, diskFileMaxSize)
        }
        this.outputAnyAdapter = builder.outputAdapter
    }

    override fun d(tag: String?, message: String?, keyword: String?) {
        val logBean = createLogBean(DEBUG, tag, message, keyword)
        printLog(logBean)
    }

    override fun e(tag: String?, message: String?, keyword: String?) {
        val logBean = createLogBean(ERROR, tag, message, keyword)
        printLog(logBean)
    }

    override fun e(tag: String?, throwable: Throwable?) {
        e(tag, throwable, null)
    }

    override fun e(tag: String?, message: String?, throwable: Throwable?) {
        e(tag, message, throwable, null)
    }


    override fun e(tag: String?, throwable: Throwable?, keyword: String??) {
        e(tag, null, throwable, keyword)
    }

    override fun e(tag: String?, message: String?, throwable: Throwable?, keyword: String?) {
        var message = message
        if (throwable != null && message != null) {
            message += " : " + Utils.getStackTraceString(throwable)
        }
        if (throwable != null && message == null) {
            message = Utils.getStackTraceString(throwable)
        }
        if (Utils.isEmpty(message)) {
            message = "Empty/NULL log message"
        }
        val logBean = createLogBean(ERROR, tag, message, keyword)
        printLog(logBean)
    }

    override fun i(tag: String?, message: String?, keyword: String?) {
        val logBean = createLogBean(INFO, tag, message, keyword)
        printLog(logBean)
    }

    override fun w(tag: String?, message: String?, keyword: String?) {
        val logBean = createLogBean(WARN, tag, message, keyword)
        printLog(logBean)
    }

    override fun v(tag: String?, message: String?, keyword: String?) {
        val logBean = createLogBean(VERBOSE, tag, message, keyword)
        printLog(logBean)
    }

    override fun d(tag: String??, message: String?) {
        val logBean = createLogBean(DEBUG, tag, message, null)
        printLog(logBean)
    }

    override fun e(tag: String??, message: String?) {
        val logBean = createLogBean(ERROR, tag, message, null)
        printLog(logBean)
    }

    override fun i(tag: String??, message: String?) {
        val logBean = createLogBean(INFO, tag, message, null)
        printLog(logBean)
    }

    override fun w(tag: String??, message: String?) {
        val logBean = createLogBean(WARN, tag, message, null)
        printLog(logBean)
    }

    override fun v(tag: String?, message: String?) {
        val logBean = createLogBean(VERBOSE, tag, message, null)
        printLog(logBean)
    }

    override fun w(message: String?) {
        w(null, message)
    }

    override fun v(message: String?) {
        v(null, message)
    }

    private fun createLogBean(debug: Int, tag: String?, message: String?, keyword: String?): LogBean {
        val logBean = LogBean()
        logBean.priority = debug
        logBean.tag = tag
        logBean.stackTraceElement = Thread.currentThread().stackTrace
        logBean.time = System.currentTimeMillis()
        logBean.content = message
        logBean.keyword = keyword
        return logBean
    }

    fun setOutputAndroidLogcat(outputAndroidLogcat: OutputAndroidLogcat) {
        this.outputAndroidLogcat = outputAndroidLogcat
    }

    fun setOutputDisk(outputDisk: OutputDisk) {
        this.outputDisk = outputDisk
    }

    fun setOutputAdapter(logAdapter: OutputAnyAdapter) {
        this.outputAnyAdapter = logAdapter
    }

    fun setLogToAndroid(logToAndroid: Boolean) {
        isLogToAndroid = logToAndroid
    }

    fun setLogToDisk(logToDisk: Boolean) {
        isLogToDisk = logToDisk
    }

    private fun printLog(logBean: LogBean) {
        handler.post {
            try {
                if (isLogToAndroid) {
                    outputAndroidLogcat?.log(logBean.copy())
                }
                if (isLogToDisk) {
                    outputDisk?.log(logBean.copy())
                }
                outputAnyAdapter?.log(logBean.copy())
            } catch (e: CloneNotSupportedException) {
                e.printStackTrace()
            }
        }
    }

    override fun d(message: String?) {
        d(null, message)
    }

    override fun e(message: String?) {
        e(null, message)
    }

    override fun i(message: String?) {
        i(null, message)
    }

    class Builder {

        var diskFilePath: String? = null
        var diskFileMaxSize: Int = 0
        var isLogToAndroid = true
        var isLogToDisk: Boolean = false

        var outputAdapter: OutputAnyAdapter? = null

        fun setDiskFilePath(diskFilePath: String): Builder {
            this.diskFilePath = diskFilePath
            return this
        }

        fun setLogOpen(isOpen: Boolean): Builder {
            isLogToAndroid = isOpen
            isLogToDisk = isOpen
            return this
        }

        fun setDiskFileMaxSize(diskFileMaxSize: Int): Builder {
            this.diskFileMaxSize = diskFileMaxSize
            return this
        }

        fun setLogToAndroid(logToAndroid: Boolean): Builder {
            isLogToAndroid = logToAndroid
            return this
        }

        fun setLogToDisk(logToDisk: Boolean): Builder {
            isLogToDisk = logToDisk
            return this
        }

        fun setOutputAdapter(outputAdapter: OutputAnyAdapter): Builder {
            this.outputAdapter = outputAdapter
            return this
        }

    }

    companion object {

        private var logPrintImpl: LogPrintImpl? = null
        /**
         * Priority constant for the println method; use Log.v.
         */
        const val VERBOSE = 2

        /**
         * Priority constant for the println method; use Log.d.
         */
        const val  DEBUG = 3

        /**
         * Priority constant for the println method; use Log.i.
         */
        const val  INFO = 4

        /**
         * Priority constant for the println method; use Log.w.
         */
        const val  WARN = 5

        /**
         * Priority constant for the println method; use Log.e.
         */
        const val  ERROR = 6

        /**
         * Priority constant for the println method.
         */
        const val  ASSERT = 7

        val instance: LogPrintImpl?
            get() {
                if (logPrintImpl == null) {
                    synchronized(LogPrintImpl::class.java) {
                        if (logPrintImpl == null) {
                            logPrintImpl = LogPrintImpl()
                        }
                    }
                }
                return logPrintImpl
            }
    }
}
